src/modem.rs

```diff
@@ -427,31 +426,167 @@ pub mod sim {
+            DBMGT51,
+            /// not known or not detectable
+            Unknown,
+            /// -116 dBm or less
+            DBMLT116,
+            /// -115 dBm
+            DBM115,
+            /// -114 to -26 dBm
+            DBM114_26(i32),
+            /// -25 dBm or greater
+            DBMGT25,
+        }
+
+        impl Display for RSSI {
+            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+                match *self {
+                    RSSI::DBMLT113 => write!(f, "<= -113 dBm"),
+                    RSSI::DBM111 => write!(f, "-111 dBm"),
+                    RSSI::DBM109_53(x) => write!(f, "{} dBm", x),
+                    RSSI::DBMGT51 => write!(f, ">= -51 dBm"),
+                    RSSI::DBM114_26(x) => write!(f, "{} dBm", x),
+                    RSSI::DBM115 => write!(f, "-115 dBm"),
+                    RSSI::DBMGT25 => write!(f, ">= -25 dBm"),
+                    RSSI::DBMLT116 => write!(f, "<= -116 dBm"),
+                    RSSI::Unknown => write!(f, "Unknown"),
+                }
+            }
+        }
+
+        impl RSSI {
+            pub fn parse(raw: i32) -> RSSI {
+                match raw {
+                    0 => Self::DBMLT113,
+                    1 => Self::DBM111,
+                    2..=30 => Self::DBM109_53(RSSI::map2_30_to_109_53(raw)),
+                    31 => Self::DBMGT51,
+                    99 => Self::Unknown,
+                    100 => Self::DBMLT116,
+                    101 => Self::DBM115,
+                    102..=191 => Self::DBM114_26(RSSI::map102_191_to_114_26(raw)),
+                    _ => Self::Unknown,
+                }
+            }
+
+            fn map2_30_to_109_53(raw: i32) -> i32 {
+                const X1: i32 = 2;
+                const Y1: i32 = -109;
+                const X2: i32 = 30;
+                const Y2: i32 = -53;
+                const GRAD: i32 = (Y2 - Y1) / (X2 - X1); // 56/28 = 2
+                const OFFSET: i32 = Y1 - (GRAD * X1); // -113
+                (GRAD * raw) + OFFSET
+            }
+
+            fn map102_191_to_114_26(raw: i32) -> i32 {
+                const X1: i32 = 102;
+                const Y1: i32 = -114;
+                // const X2: i32 = 191;
+                // const Y2: i32 = -26;
+                const GRAD: i32 = 1;
+                // requires #![feature(int_roundings)]
+                // const GRAD: i32 = (Y2 - Y1).div_ceil((X2 - X1)); // would be 88/89, so truncated to 0
+                const OFFSET: i32 = Y1 - (GRAD * X1); // -216
+                (GRAD * raw) + OFFSET
+            }
+        }
+
         impl Default for SIM7600 {
             fn default() -> Self {
                 Self::new()
             }
         }
 
         impl SimModem for SIM7600 {
-            fn negotiate(&mut self, comm: &mut UartDriver) -> Result<(), ModemError> {
-                reset(comm)?;
+            fn negotiate(
+                &mut self,
+                comm: &mut UartDriver,
```

Comment:
You are returning the `UartDriver` where you already had a nice abstraction just because you need to do a read-with-a-timeout.

This is unnecessary. Reading with a timeout is as simple as combining `Read` with - say - `embedded_hal_async::Delay` like so:

```rust
async fn read_with_timeout<R: embedded_io_async::Read, D: embedded_hal_async::DelayNs>(read: R, delay: D, buf: &mut [u8], timeout_ms: u16) -> Result<usize, R::Error> {
    let read = read.read(buf);
    let timeout = delay.delay_ms(timeout_ms as _);

    match embassy_futures::select::select(read, timeout).await {
        Either::First(result) => result,
        Either::Second(_) => Ok(0), // 0 bytes read indicates a timeout
    }
}
```

Reply:
@DaneSlattery  Ah sorry!
I did not realize your code still lives in "blocking"-land and not in `async`-land yet. :D Perhaps we can delay the re-introduction of `embedded_io_async::Read + Write` until after we start to switch this code from blocking to async! For now, maybe just make it work with `UartDriver` (where the you-assume-that-your-read-reply-will-contain-the-whole-response issue is _the_ more important issue), and once we have something mostly working reliably, we can switch from blocking to async. And then still layer blocking _on top_ of async (Or use `AsyncUartDriver` maybe).

Reply:
(
The issue at hand is that stuff like `read_with_timeout` is much, much more difficult to implement in blocking-land (would require scoped threads) than it is in async-land. This is one place where `async` is strictly more powerful than blocking - the "intra-task concurrency" from the blog of Boats, if you recall that.
)

Reply:
I saw your comments on the PR, and I was only using the timeout to force UartDriver:::read to return the bytes. I was confident that after 1s the modem will have replied, but that is completely arbritrary. I like the solution to feed parts into a Digester that is a state machine searching for some terminator (like '\r\nOK\r\n'.as_bytes(). This can happen after a few read cycles. After the terminator is found, then we can try parse it using the at_commands parser. If the terminator is not found, the digester will keep waiting. It would actually be really useful if at_commands::parser could provide expects_terminator(...) and throw a different error if the terminator is not found.